File: Python3/807. Max Increase to Keep City Skyline.py
class Solution:
    def maxIncreaseKeepingSkyline(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        row = len(grid)
        col = len(grid[0])
        left = [0] * row
        top = [0] * col
        for i in range(row):
            for j in range(col):
                if grid[i][j] > top[j]:
                    top[j] = grid[i][j]
                if grid[i][j] > left[i]:
                    left[i] = grid[i][j]
        count = 0
        for i in range(row):
            for j in range(col):
                count += min(left[i], top[j]) - grid[i][j]
        return count

File: Golang/807. Max Increase to Keep City Skyline.go
func maxIncreaseKeepingSkyline(grid [][]int) int {
    row := len(grid)
    col := len(grid[0])
    left := make([]int, row)
    top := make([]int, col)
    for i := 0; i < row; i++ {
        for j := 0; j < col; j++ {
            if grid[i][j] > top[j] {
                top[j] = grid[i][j]
            }
            if grid[i][j] > left[i] {
                left[i] = grid[i][j]
            }
        }
    }
    count := 0
    for i := 0; i < row; i++ {
        for j := 0; j < col; j++ {
            count += min(left[i], top[j]) - grid[i][j]
        }
    }
    return count
}

func min(a, b int) int {
    if a > b {
        return b
    } else {
        return a
    }
}
